### PR TITLE
Improve displaying info about insufficient stock in the cart

### PR DESCRIPTION
**Pull request by jakubtobiasz**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13
| Bug fix?        | no
| New feature?    | kinda
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | improvements based on #14865
| License         | MIT

There are few changes not suitable to be pushed as "bug fix" in `1.12`. 

Preview:
CleanShot 2023-03-21 at 13 53 22@2x

### CODE REVIEW

File: src/Sylius/Bundle/InventoryBundle/Validator/Constraints/InStockValidator.php [hunk 1]
```diff
@@ -22,6 +23,10 @@
 
 final class InStockValidator extends ConstraintValidator
 {
+    private const TARGET_CLASS = 'class';
+
+    private const TARGET_PROPERTY = 'property';
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/InventoryBundle/Validator/Constraints/InStockValidator.php]: Instead of those, the constants from `Symfony\Component\Validator\Constraint` should be used.

File: .github/workflows/ci_e2e-custom.yaml [hunk 2]
```diff
@@ -60,5 +60,7 @@ jobs:
                 if: failure()
                 with:
                     name: "Behat logs (non-JS, PHP ${{ matrix.php }}, Symfony ${{ matrix.symfony }}) [Unstable], MySQL ${{ matrix.mysql }})"
-                    path: etc/build/
+                    path: |
+                        etc/build/
+                        var/log
```
Comment (NoResponseMate) [inline .github/workflows/ci_e2e-custom.yaml]: Should go in a different PR

File: src/Sylius/Behat/Page/Shop/Cart/SummaryPage.php [hunk 3]
```diff
@@ -185,6 +185,13 @@ public function hasItemWithCode(string $code): bool
         return $this->hasItemWith($code, '[data-test-product-variant-code]');
     }
 
+    public function hasProductWithInsufficientStock(ProductInterface $product): bool
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Page/Shop/Cart/SummaryPage.php]:
```suggestion
    public function hasItemWithInsufficientStock(string $productName): bool
```
To be consistent

File: features/inventory/cart_inventory/verify_inventory_quantity_when_updating_cart_summary.feature [hunk 4]
```diff
@@ -9,13 +9,27 @@ Feature: Verifying inventory quantity on cart summary
...
         Given I have added 3 products "Iron Maiden T-Shirt" in the cart
         When I change product "Iron Maiden T-Shirt" quantity to 6 in my cart
         And I update my cart
-        Then I should be notified that this product cannot be updated
+        Then I should be notified that this product has insufficient stock
+
+    @ui @no-api
+    Scenario: Preventing the cart recalculation when the form has errors
+        Given I have added 3 products "Iron Maiden T-Shirt" in the cart
+        And I have added 1 products "Black Dress" in the cart
```
Comment (GSadee) [inline features/inventory/cart_inventory/verify_inventory_quantity_when_updating_cart_summary.feature]:
```suggestion
        And I have added a single product "Black Dress" to the cart
```

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation/OrderItem.xml [hunk 5]
```diff
@@ -13,11 +13,18 @@
 
 <constraint-mapping xmlns="http://symfony.com/schema/dic/constraint-mapping" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/constraint-mapping http://symfony.com/schema/dic/services/constraint-mapping-1.0.xsd">
     <class name="Sylius\Component\Order\Model\OrderItem">
+        <getter property="quantity">
+            <constraint name="Sylius\Bundle\InventoryBundle\Validator\Constraints\InStock">
+                <option name="stockablePath">variant</option>-->
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Resources/config/validation/OrderItem.xml]:
```suggestion
                <option name="stockablePath">variant</option>
```

File: src/Sylius/Bundle/InventoryBundle/Validator/Constraints/InStockValidator.php [hunk 6]
```diff
@@ -34,21 +39,31 @@ public function validate($value, Constraint $constraint): void
...
+        }
+
+        if (self::TARGET_CLASS === $target) {
             $this->context->addViolation(
                 $constraint->message,
                 ['%itemName%' => $stockable->getInventoryName()],
             );
+        } else {
+            $this->context->addViolation($constraint->shortMessage);
         }
```
Comment (GSadee) [inline src/Sylius/Bundle/InventoryBundle/Validator/Constraints/InStockValidator.php]:
```suggestion
        if (self::TARGET_CLASS === $target) {
            $this->context->addViolation(
                $constraint->message,
                ['%itemName%' => $stockable->getInventoryName()],
            );
            
            return;
        }
        
        $this->context->addViolation($constraint->shortMessage);
```

File: src/Sylius/Bundle/InventoryBundle/spec/Validator/Constraints/InStockValidatorSpec.php [hunk 7]
```diff
@@ -44,6 +48,19 @@ function it_does_not_add_violation_if_there_is_no_stockable(
...
+        InventoryUnitInterface $inventoryUnit,
+        PropertyAccessor $propertyAccessor,
+        ExecutionContext $context,
+    ): void {
+        $context->getObject()->willReturn($inventoryUnit);
+        $propertyAccessor->getValue($inventoryUnit, 'stockable')->willReturn(null);
+
+        $constraint = new InStock();
+
+        $this->validate(1, $constraint);
```
Comment (GSadee) [inline src/Sylius/Bundle/InventoryBundle/spec/Validator/Constraints/InStockValidatorSpec.php]:
```suggestion
        $this->validate(1, new InStock());
```

File: features/inventory/cart_inventory/verify_inventory_quantity_when_updating_cart_summary.feature (hunk 4 above)
Comment (jakubtobiasz) [reply features/inventory/cart_inventory/verify_inventory_quantity_when_updating_cart_summary.feature]: I left "I have added 1 product "Black Dress" to the cart".

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Jacob! :tada:
